### PR TITLE
Add Order model

### PR DESCRIPTION
**Pull request by dmarmul**
Added Order and OrderItem to project. Implement all needed classes for it.

### CODE REVIEW

File: src/main/java/org/example/bookshop/controller/CartController.java [hunk 1]
```diff
@@ -41,7 +41,7 @@ public ShoppingCartDto get(@AuthenticationPrincipal User user) {
     @PreAuthorize("hasRole('USER')")
     @Operation(summary = "Add books to cart",
             description = "Add books to the shopping cart."
-                    + " Fields bookId and quantity can't must be not less than 1, can't be null.")
+                    + " Fields bookId and quantity must be positive, can't be null.")
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/controller/CartController.java]: in programming, 0 is considered a positive number)

File: src/main/java/org/example/bookshop/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public OrderDto add(@RequestBody @Valid OrderPurchaseRequestDto requestDto,
+                        @AuthenticationPrincipal User user) {
+        return orderService.add(requestDto, user);
+    }
+
+    @GetMapping
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Get orders history",
+            description = "Retrieve all user's orders")
+    public Set<OrderDto> getAll(@AuthenticationPrincipal User user,
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/controller/OrderController.java]: better return Page

File: src/main/java/org/example/bookshop/service/OrderService.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import org.example.bookshop.model.User;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+
+public interface OrderService {
+    OrderDto add(@Valid OrderPurchaseRequestDto requestDto, User user);
+
+    Set<OrderDto> getAll(User user, Sort sort, Pageable pageable);
+
+    void update(@Valid OrderStatusRequestDto requestDto, Long orderId, User user);
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/OrderService.java]: validation on the controller layer is enough

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                user.getId()).orElseThrow(() -> new EntityNotFoundException(
+                "Can't find order for user: " + user.getEmail()));
+        Order order = orderMapper.toEntity(shoppingCart);
+        order.setShippingAddress(requestDto.getShippingAddress());
+        order.setOrderDate(LocalDateTime.now());
+        order.setStatus(Order.Status.PENDING);
+        Set<OrderItem> orderItems = findOrderItems(user);
+        orderItems.forEach(item ->
+                item.setPrice(item.getPrice().multiply(new BigDecimal(item.getQuantity()))));
+        order.setTotal(getTotalPrice(orderItems));
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]: better move this logic to the private method

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+public class OrderServiceImpl implements OrderService {
+    private final CartItemRepository cartItemRepository;
+    private final CartRepository cartRepository;
+    private final OrderRepository orderRepository;
+    private final OrderItemRepository orderItemRepository;
+    private final OrderItemMapper orderItemMapper;
+    private final OrderMapper orderMapper;
+
+    @Override
+    public OrderDto add(OrderPurchaseRequestDto requestDto, User user) {
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]: Let’s  use Transactional here ;)

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public void update(OrderStatusRequestDto requestDto, Long orderId, User user) {
+        Order order = findOrderByIdAndUserId(orderId, user.getId());
+        Order.Status status = Order.Status.valueOf(requestDto.getStatus());
+        if (order.getStatus().equals(status)) {
+            throw new DataProcessingException("Order already has status: " + status);
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:64]: I don’t see your global exception handler but I guess it provide internal server error. is it ok?

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+                .map(orderItemMapper::toDto)
+                .collect(Collectors.toSet());
+        // Have alternative solve, it should be faster,
+        // but need 1 more method in orderItemRepository
+        // return orderItemRepository.findAllByOrderId(orderId)
+//                .stream()
+//                .filter(Optional::isPresent)
+//                .map(Optional::get)
+//                .map(orderItemMapper::toDto)
+//                .collect(Collectors.toSet());
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]: Let’s  remove the commented code

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        return orderItemMapper.toDto(itemOptional.get());
+    }
+
+    private Order findOrderByIdAndUserId(Long orderId, Long userId) {
+        Optional<Order> optionalOrder = orderRepository.findByIdAndUserId(orderId, userId);
+        if (optionalOrder.isEmpty()) {
+            throw new EntityNotFoundException("Can't find order with id: " + orderId);
+        }
+        return optionalOrder.get();
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]:
```suggestion
    private Order findOrderByIdAndUserId(Long orderId, Long userId) {
        return orderRepository.findByIdAndUserId(orderId, userId).orElseThrow(() -> 
                 new EntityNotFoundException("Can't find order with id: " + orderId)
        );
    }
```

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+//                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public OrderItemDto getOrderItem(Long orderId, Long itemId, User user) {
+        findOrderByIdAndUserId(orderId, user.getId());
+        Optional<OrderItem> itemOptional = orderItemRepository.findByIdAndOrderId(itemId, orderId);
+        if (itemOptional.isEmpty()) {
+            throw new EntityNotFoundException("Can't find item with id: " + itemId);
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]: Let’s  use `orElseThrow`

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    private Set<OrderItem> findOrderItems(User user) {
+        Set<OrderItem> orderItem = cartItemRepository.findAllByShoppingCartId(
+                cartRepository.findByUserId(user.getId()).get().getId())
+                .stream()
+                .map(orderItemMapper::toEntity)
+                .collect(Collectors.toSet());
+        if (orderItem.isEmpty()) {
+            throw new EntityNotFoundException(
+                    "Can't find order items for user: " + user.getEmail());
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]: why should it occur Exception?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java (hunk 6 above)
Comment (dmarmul) [reply src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:64]: There is provides 401 http response

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java (hunk 10 above)
Comment (dmarmul) [reply src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]: This check doesn't allow to place an empty order. Added more description into exception

File: src/main/java/org/example/bookshop/controller/OrderController.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    public Page<OrderDto> getAll(@AuthenticationPrincipal User user,
+                                 Sort sort, Pageable pageable) {
+        return orderService.getAll(user, sort, pageable);
+    }
+
+    @PatchMapping("/{orderId}")
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Update order status",
+            description = "Update order status")
+    public void update(@RequestBody @Valid OrderStatusRequestDto requestDto,
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/controller/OrderController.java]:
```suggestion
    public OrderDto update(@RequestBody @Valid OrderStatusRequestDto requestDto,
```

File: src/main/java/org/example/bookshop/controller/OrderController.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+                       @PathVariable Long orderId,
+                       @AuthenticationPrincipal User user) {
+        orderService.update(requestDto, orderId, user);
+    }
+
+    @GetMapping("/{orderId}/items")
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Get OrderItems",
+            description = "Retrieve all OrderItems for a specific order")
+    public Set<OrderItemDto> get(@PathVariable Long orderId,
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/controller/OrderController.java]: rename method

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    private final CartItemRepository cartItemRepository;
+    private final CartRepository cartRepository;
+    private final OrderRepository orderRepository;
+    private final OrderItemRepository orderItemRepository;
+    private final OrderItemMapper orderItemMapper;
+    private final OrderMapper orderMapper;
+
+    @Override
+    @Transactional
+    public OrderDto add(OrderPurchaseRequestDto requestDto, User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]: rename method

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+            throw new DataProcessingException("Order already has status: " + status);
+        }
+        order.setStatus(status);
+        orderRepository.save(order);
+    }
+
+    @Override
+    public Set<OrderItemDto> get(Long orderId, User user) {
+        findOrderByIdAndUserId(orderId, user.getId());
+        return orderItemRepository.findAllByOrderId(orderId).stream()
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java]: Don't need to request DB twice, 
see an example
```suggestion
        Order order = orderRepository.findByIdAndUserId(orderId, userId)....
        return orderItemMapper.toOrderItemDtoList(order.getOrderItems());
```

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+                .map(orderItemMapper::toDto)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public OrderItemDto getOrderItem(Long orderId, Long itemId, User user) {
+        return orderItemMapper.toDto(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:78]:
```suggestion
         OrderItem item = itemRepository.findByIdAndOrderIdAndUserId(orderItemId, orderId, userId)
                .orElseThrow(()....
         return orderItemMapper.toOrderItemDto(item);
```

File: src/main/java/org/example/bookshop/model/Order.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private Order.Status status;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/model/Order.java]: you can initialize status here

File: src/main/java/org/example/bookshop/model/Order.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/model/Order.java]: you can initialize date now() here

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java (hunk 6 above)
Comment (liliia-ponomarenko) [reply src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:64]: Okay, your filter chain does it. Anyway, you see an exception in the console. Is it ok? 
401 Unauthorized  - is it the appropriate status for this exception? 
You should handle this exception and return CONFLICT.

**Review by liliia-ponomarenko — APPROVED** [review]
Great! But please pay attention to my comment ;)

**Pull request merged**